pkg/recommendation/device.go

```diff
@@ -0,0 +1,235 @@
+package recommendation
+
+import (
+	"sort"
+
+	"github.com/pkg/errors"
+	"k8s.io/apimachinery/pkg/api/resource"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+
+	"mayadata.io/cstorpoolauto/types"
+	bdutil "mayadata.io/cstorpoolauto/util/blockdevice"
+)
+
+// nodeCapacityBlockDevice contains key with node name and value with capacityBlockDevice
+type nodeCapacityBlockDevice map[string]capacityBlockDevices
+
+// capacityBlockDevices contains a key with capacity of block device and
+// value with all the block devices of that capacity.
+type capacityBlockDevices map[int64][]bdutil.MetaInfo
+
+type cStorPoolClusterRecommendationRequest types.CStorPoolClusterRecommendationRequest
+
+// NewDeviceRequest returns a device request object after validation.
+func NewDeviceRequest(
```

Comment:
Will it be good if client code looks has following:
```go
recommendation.RequestByDevice(typecspc.RecommendationData)
//vs.
recommendation.NewDeviceRequest(...)
```

where typecspc.RecommendationData has all the PUBLIC fields required to build a typecspc.RecommendationRequest instance.

Reply:
@shovanmaity please take a look. We can do them later.
However, lets ensure there is no breakages due to later changes.

Reply:
server: 
```
// Data is the input data requested for the recommendation.
type Data struct {
	RaidConfig      *types.RaidGroupConfig
	BlockDeviceList *unstructured.UnstructuredList
	PoolCapacity    *resource.Quantity
}

// NewRequestForDevice returns a device request object after validation.
func NewRequestForDevice(data Data) (
	*cStorPoolClusterRecommendationRequest, error) {...}
```
client:
```
inputData := recommendation.Data{
	RaidConfig:      raidconfig,
	BlockDeviceList: list,
	PoolCapacity:    &poolCapacity,
}

r, err := recommendation.NewRequestForDevice(inputData)
```

Is this better? @AmitKumarDas